Time: add setitem and missing value support (masking)

taldcroft opened the pull request:
This is an implementation of the necessary bits to fully support table join, hstack, and vstack for the Time class.

It includes new functionality for Time masking and setting, along with an implementation of the info class `new_like()` method.

**Masking**: there was discussion about whether to do this with `np.nan` internally or using `MaskedArray`.  I first did the latter and got something mostly working (passing most but not all tests), but it ended up getting messy.  In particular the interactions with `Quantity` got ugly because Quantity knows how to play with `ndarray` but not `MaskedArray` in arithmetic etc.

But it looks like the `np.nan` solution is fine, with the unexpected twist that using this strictly in `jd2` makes life a lot easier because it is easy to locally set the `nan` values to `0` before entering any ERFA functions.  This avoids the `dubious` values problem.

**Setting**: I think it is time to make `Time` mutable!  Admittedly I haven't thought super hard about pitfalls yet, but I think just being careful about clearing the cache is enough.  This is going to be needed to make `Time` work in table operations (and probably Time series) where you need masking and item setting.

To do:

- [x] Tests of setting
- [x] Docs
- [x] Changelog

mhvk commented on astropy/time/formats.py:
```diff
@@ -144,11 +150,40 @@ def scale(self):
+
+    @property
+    def jd2_filled(self):
+        return np.nan_to_num(self.jd2)
```
to ensure that the regular no-nan case is not slowed down too much:
```
return np.nan_to_num(self.jd2) if self.masked else self.jd2
```

mhvk replied:
to ensure that the regular no-nan case is not slowed down too much:
```
return np.nan_to_num(self.jd2) if self.masked else self.jd2
```

taldcroft replied:
Yup, I had also thought about that.

mhvk commented on astropy/time/core.py:
```diff
@@ -583,6 +588,40 @@ def value(self):
+                except Exception:
+                    raise ValueError('cannot convert value to a compatible Time object')
+
+        if self.location != value.location:
```
`location` can be an array, in which case this statement would break. Now it is guaranteed that location is either a single element or an array with the correct shape, but the shape may have been broadcast, so one cannot just set an element. So, this needs some trickery.... But it also means in principle a different location is fine. Similarly, if `value.location is None`, I think it might be treated as OK if `self.location is not None` (but arguably only if `self.location` is a single element).

taldcroft replied:
I think that a non-scalar location is a corner case and that we could just raise an exception if either the `self` or `value` location is not a scalar or `None`.  At least for now this would be fine and we can see if any users actually complain.

mhvk commented on astropy/time/formats.py:
```diff
@@ -72,6 +73,11 @@ def _regexify_subfmts(subfmts):
     return tuple(new_subfmts)
 
 
+class TimeMaskedArray(np.ma.MaskedArray):
```
Why do we need a new class? It seems you want to avoid `repr`, but where is that a problem?

taldcroft replied:
Otherwise you get this:
```
In [9]: tm.cxcsec
Out[9]: 
masked_array(data = [99.99999999999805 -- 299.999999999999],
             mask = [False  True False],
       fill_value = 1e+20)
```
Instead of:
```
In [4]: tm.cxcsec
Out[4]: [99.99999999999805 -- 299.999999999999]
```

mhvk replied:
But I think that's OK. Right now, if I don't mask anything, I get:
```
In [23]: Time(np.arange(50000, 50010), format='mjd').cxcsec
Out[23]: 
array([-70329538.816, -70243138.816, -70156738.816, -70070338.816,
       -69983938.816, -69897538.816, -69811138.816, -69724738.816,
       -69638338.816, -69551938.816])
```
so why shouldn't a masked `Time` return a `MaskedArray`?

taldcroft replied:
I don't like the `masked_array` repr because:
- I don't need to see an explicit `data = ..`, which has `--` marking masked elements, AND `mask = [..]`, which is basically the same information.
- The `fill_value` is just a silly thing to include.  It is irrelevant *most* of the time.
- The `data = ` is redundant.  I know it is the data.

One can say this uses an old-school definition of repr, that it gives you the info to reproduce the object (although there it fails because it doesn't include dtype). But I think in this age of Jupyter notebook / interactive analysis, it is more important to have the object repr be a concise and visually-friendly representation of the object.

taldcroft replied:
For consistency it could be:
```
def __repr__(self):
    return 'masked_array({})'.format(self)
```

taldcroft replied:
OK, got rid of the TimeMaskedArray wrapper.

mhvk commented:
For `location`, this is a `Quantity`, so in principle one can set elements already. I think one needs a function something like the following in `__setitem__`:
```
if self.location is None:
    if value.location is not None:
        # Cannot have some locations unknown and some set.
        raise  
    # no locations involved; just go on
else:
    if value.location is None:
        if self.location.size > 1:
            # have different locations, but nothing set
            raise
        # value does not set location, so keep the single one we have
    else:
        if np.any(self.location[item] != value.location):
            # remove any broadcasting
            if not self.location.flags.writable:
                new_location = self.location.copy()
            new_location[item] = value.location
            # maybe do it at end, so `self` remains unchanged if there are errors.
            self.location = new_location
```

Now the above is obviously a bit elaborate, but while we perhaps could get rid of broadcasting, I think we should continue to allow single locations, especially as we will need to do the same for `SkyCoord` and its many attributes anyway... Which, indeed, suggests this function could live on `ShapedLikeNDArray` (which already implements `__getitem__`).

Now, *given* the need for dealing with attributes that do not necessarily share the objects shape, I think one way to make `mask` easier is to treat it as such an attribute as well, i.e., it can be `None` or a (partially broadcast) array (a single item seems not so useful in this case).

Also a suggestion: I think at some level setting items and having a mask are different issues; would it be an idea to decouple the two? If we do `__setitem__` first, and ensure it is tested thoroughly, we will be in a better place to see if my idea of treating `mask` like another attribute will work or not.

mhvk commented:
p.s. Obviously, just having `__setitem__` will at least allow `vstack` of tables, which probably is a main use.  And it would allow defining `insert` as well...

For the mask, I think one decision we need to make is the extent to which this is a work-around for table, or general. E.g., you return masked array for all the properties, but what about `time_delta.to(u.s)`?

taldcroft commented:
> For the mask, I think one decision we need to make is the extent to which this is a work-around for table, or general. E.g., you return masked array for all the properties, but what about time_delta.to(u.s)?

My thinking is that this is general.  Time properties (in particular all format properties) will be a masked array.  There is a question about things like `delta_ut1_utc`.  That is not currently masked, but I guess it could be put through `mask_if_needed`.

Any Quantity output will just be a Quantity with `nan`:
```
In [8]: dt
Out[8]: <TimeDelta object: scale='tt' format='jd' value=[0.001041666666666663 -- 0.003124999999999989]>

In [9]: dt.to(u.s)
Out[9]: <Quantity [  90.,  nan, 270.] s>
```
In case you haven't guessed, I'm pushing for `nan` as the formal indicator of masking for Quantity.  It's something we can easily do for 2.0.  I admit I always thought Quantity had to be float until you mentioned it recently, so this is not a 100% solution.  But for this (and probably 98% or more) of use cases, this is fine.

taldcroft commented:
I think one principle that will keep the Time masking implementation manageable and simple is to *not* apply the mask to all non-format attributes, e.g. location.  As it currently stands `mask` is a derived attribute which is necessarily the same shape as the data.  I think that trying to shoehorn this into every other Time attribute, which may or may not share the shape is going to make this much messier and potentially introduce unexpected changes to users.

taldcroft commented:
About splitting this into two PRs, in principle I agree that is the right thing.  In practice it is going to make things go faster for me if I don't.  I have a limited amount of time before the day-job pulls me back and substantially slows development.  In fact I changed the title and you'll see that I'm about to include what could reasonably be a *3rd* PR.  Note that the 3rd would require 1 (setting) and 2 (masking), and 2 would require 1, so you can see how developing them independently just slows the process.

mhvk commented:
@taldcroft - I'm much more uncomfortable with the mask than with `__setitem__`, as the particular way of doing the mask has implications through other parts of astropy (`Quantity`, clearly, which may still more logically be done with `MaskedArray`, but also `SkyCoord`, where the trick of just doing `jd2` is not available).

Perhaps more importantly, just having `__setitem__` addresses at least one large table problem (the largest?) -- that `vstack` and `insert` are not possible. And, as said, I can also see how it nearly trivially extends to `SkyCoord` as long as we treat attributes consistently from the get-go.

mhvk commented:
p.s. I do worry quite a bit about unexpected consequences of making `Time` mutable. As is, internal arrays can be shared between instances, so if one changes, another will too, but all code written will have implicitly assumed this never happens. We may want a `mutable` flag which by default is `False`...

taldcroft commented:
> as the particular way of doing the mask has implications through other parts of astropy

From my perspective this is completely internal and the fact that `_time.jd2` has NaN's is not exposed at all (at least that is the intent).  So we would be free to change the internal implementation in the future at any time.  

Now it is true that in this implementation, a `Quantity` output like the `.to(..)` method assumes that `Quantity` implements masking by the NaN sentinel.  We could certainly hold off on this decision by making `.to(..)` etc raise an exception for a masked Time object.  I'm OK with that, but I will say:

- I think that you are underestimating the difficulty of making a `MaskedContainer` or `MaskedQuantity`.  My gut feeling is that at the end it will be a *lot* of work to develop and maintain.
- For the case of float arrays, I think that using NaN is just the better way.  Everything is already built-in and it is fast and memory-efficient.

> Perhaps more importantly, just having __setitem__ addresses at least one large table problem (the largest?) -- that vstack and insert are not possible.

I'm tired of mixins being crippled.  By adopting the NaN protocol they can be fully functional!  I'm mostly thinking of join(), which is actually very useful and (except for inner) needs masking.

taldcroft commented:
>  SkyCoord, where the trick of just doing jd2 is not available

True, but setting the representation values to `nan` could work, though currently this generates warnings from the Angle classes.  Admittedly this is a bigger beast than Time.

taldcroft commented:
About changing Time to be mutable, I have sent email to astropy-dev to request input.  I agree on the idea of some attribute (`writeable` like the numpy flag?) that can be used to prevent setting.  However I would have that default to `writeable = True`.

In general I think developers are very much used to the pitfalls of shared memory in numpy array objects, so I think we should enable writing by default.  Since no existing code writes to Time objects, on the transition to 2.0 nothing would immediately break.  For the small minority of cases where mutability is a problem, developers would have the opportunity to fix their code.  

This could be as simple (and back-compatible) as `t.writeable = False`.  For pre-2.0 this would just set an attribute that is not used anywhere.  For post-2.0 this could be a property that is a view of the `flags.WRITEABLE` property of the underlying `_time.jd1,2`.  (That's just an idea, haven't thought it through all the way, but the goal is to avoid a new Time attribute that needs to be manually managed).

dkirkby commented:
I like the idea of mirroring the numpy [WRITEABLE flag](https://docs.scipy.org/doc/numpy-1.12.0/reference/generated/numpy.ndarray.flags.html) since that allows new functionality to be implemented gradually but probably also requires that the default is `WRITEABLE = False`, at least initially.

An alternative paradigm would be to base a new `MutableTime` class off `Time`, but I would only recommend that in a strongly typed language.

mhvk commented:
@eteq - this is the PR for making `Time` (somewhat) mutable -- as you expressed a strong opinion about possibly doing the same for coordinates, now would be the moment to think about an API that would make sense (see discussion above; part of this is needed to be able to merge to sets of times -- the same would be useful for coordinates; for both, this is already somewhat in place for merging a list of scalars).

taldcroft commented:
I've been working on this but I see that making it for 2.0 is perhaps not likely at this point.

bsipocz commented:
@taldcroft - I'm a bit confused regarding this PR. Should we wait for it in the RC (there are still a few outstanding coordinates PR, so there is still some time to wrap up), or it's already been moved to 3.0? If the latter, could you update the milestone?

taldcroft commented:
@bsipocz - sorry, this has been moved to 3.0 and re-milestoned accordingly.

Cadair commented:
@taldcroft Sorry if I have not understood from the previous discussion, but would this PR allow the reordering of a Time column? (Something else that will be needed for timseries tables).

taldcroft commented:
> would this PR allow the reordering of a Time column?

Sure!
```
In [2]: ii = np.arange(5)
In [3]: t = Time(ii, format='cxcsec')
In [4]: np.random.shuffle(ii)
In [6]: ii
Out[6]: array([0, 4, 1, 2, 3])

In [7]: t
Out[7]: <Time object: scale='tt' format='cxcsec' value=[ 0.  1.  2.  3.  4.]>

In [8]: t[:] = t[ii]

In [9]: t
Out[9]: <Time object: scale='tt' format='cxcsec' value=[ 0.  4.  1.  2.  3.]>
```
Likewise if you have a time column in a table then the same will work to re-order the entire table.

mhvk commented:
And time already has an `argsort()` to help this. Arguably, we could now add an option to `sort` to do the sorting in-place.

astrofrog commented on astropy/time/core.py:
```diff
@@ -990,7 +1136,14 @@ def argmin(self, axis=None, out=None):
+            if axis is not None:
+                approx = np.expand_dims(approx, axis)
+        else:
+            approx = np.min(jd, axis, keepdims=True)
```
@taldcroft - I think the issue is here for the CircleCI issue. On CircleCI we have Numpy 1.11, and it looks like the last line doesn't actually preserve dimensions:

```
(Pdb) np.__version__
'1.11.0'
(Pdb) np.min(jd, axis, keepdims=True)
masked_array(data =
 [[2450000.6 -- 2450000.6 2450001.4 2450001.4]
 [2450050.6 -- 2450050.6 2450051.4 2450051.4]],
             mask =
 [[False  True False False False]
 [False  True False False False]],
       fill_value = 1e+20)

(Pdb) np.min(jd, axis, keepdims=True).shape
(2, 5)
```

Should ``NUMPY_LT_1_11`` be ``NUMPY_LE_1_11`` or ``NUMPY_LT_1_12``?

astrofrog replied:
@mhvk - is this related to the fix that was included in 1.11.2? (https://github.com/numpy/numpy/pull/8018) If so then I guess this should be NUMPY_LT_1_11_2.

taldcroft replied:
@astrofrog - ah right.  When I was testing locally I just tried numpy 1.11 which was the latest bugfix version.  Will fix.

mhvk replied:
@astrofrog - nice sleuthing!

mhvk commented on astropy/time/core.py:
```diff
@@ -1043,6 +1205,7 @@ def min(self, axis=None, out=None, keepdims=False):
         ``np.min``; since `Time` instances are immutable, it is not possible
         to have an actual ``out`` to store the result in.
         """
+
```
Bit of a drive-by comment, but why the extra blank lines here and below? Blank lines after docstrings are only suggested for modules and classes by PEP 257 - https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

taldcroft replied:
No strong reason.  I've been influenced a bit by the [plenty-of-vertical-whitespace style](https://github.com/astrofrog/fast-histogram/blob/master/fast_histogram/histogram.py) of @astrofrog, which often includes a blank line after the docstring.  Also it somehow feels just a little easier to see the first code line after a multiline docstring with that space.  I don't quite follow the distinction that PEP-257 makes between functions and class methods.

saimn replied:
On the [Numpy doc guide](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) the blank line is always before the quotes, not after. So that's what I use in my projects, but not sure if there is a real recommendation about this.

larrybradley replied:
I include a blank line after docstrings (for both classes and functions; IMHO it makes things more readable).  Several people put a blank line before the end quotes, but I've never understood the reason for that or knew where it came from (it's not in PEP-257).

mhvk replied:
I always felt the closing three quotes on a line make for a clear separation - and there is also a benefit to having more code on a single screen-full.

In any case, it probably ends up a matter of taste, so I'd say that we follow the general rule to not stick with what a previous person put there unless it clearly violates pep8/pep257 (and definitely not change away from pep8/pep257).

(I write this trying to restrain myself every time I see lines that are longer than 80 characters and which thus wrap in my editor).

taldcroft replied:
I agree fully with not changing the style of existing code and minimizing changes.  Like many public figures these days, I don't actually remember doing this.  But I do have the option of reverting this past mistake and will do so.  :smile:

taldcroft commented:
@mhvk - I split out the table operations stuff (join, hstack, vstack) to #6888, which is based off this branch.  The exception is b963ae4 because it join breaks if a mixin column has masking but the `mask` attribute is not directly settable.  (I *think* that is what was going on, but when I took out this commit then the table inner join test failed).

taldcroft commented:
@mhvk - 3 weeks to feature freeze.  Will you be able to look at this fairly soon?

taldcroft commented:
Ready for final review.

mhvk commented on astropy/time/tests/test_methods.py:
```diff
@@ -280,21 +293,29 @@ def test_broadcast(self):
     """Arithmetic on Time objects, using both doubles."""
-    kwargs = ({}, {'axis': None}, {'axis': 0}, {'axis': 1}, {'axis': 2})
-    functions = ('min', 'max', 'sort')
+    # kwargs = ({}, {'axis': None}, {'axis': 0}, {'axis': 1}, {'axis': 2})
```
Forgot to "comment in"?

taldcroft replied:
Yup, fixed.

mhvk commented:
@taldcroft - I now had somewhat more detailed look. First, I remain not sold on making `Time` mutable, and think that should be discussed separately, together with whether we want to (eventually) move that way for `SkyCoord`. Overall, my feeling is that abandoning immutability is in the end going to give much increased complexity and hurt performance. I would like to understand the benefits we gain better, especially if it was just `Time` that became mutable.  (Just as one example of something that currently goes wrong: `_delta_utc_ut1` is not propagated. So, one could end up with wrong corrections. Obviously, one can just delete them, but that removes also anything set by hand.)

I can see the benefits of having a mask more easily, though again a worry is performance and complexity: over time, there are going to be more and more corner cases where `MaskedArray` doesn't quite do what we want and we have to special-case stuff. But the PR does seem encouraging in it not being that bad.

I do wonder about the implementation, and in particular about the changes to the format classes (I have a `TimeGSB` class myself...). I would have thought it would be cleanest to either use `MaskedArray` directly for the internal `jd1` and `jd2`, or to do all mask-related stuff on the `Time` class. Is the former tricky mostly given the erfa routines? If so, would it help if one ensured that any masked value would be a valid time?

And if one doesn't use `MaskedArray`, wouldn't it make more sense to keep the format classes exactly as they are, and apply/remove masks in `Time` itself? It would seem this would be possible. More generally, I guess I don't quite understand the use of setting jd2 to `nan`. Originally, I think the idea was that it would mean masks would propagate automatically, but in the end it seems that any time it gets tricky, you need to use `jd2_filled` anyway, and then recreate the mask. A mask that is a lighter shim between inputs and outputs might be a better solution (in which case a mask would also safely be writable).

A separate issue, not easily solved without `MaskedQuantity`, is that the interaction between `Time` and quantities now becomes trickier.

p.s. I should add that the work with @eteq on the `Distribution` class (#6945) has given me hope that adding our own `Masked[Anything]` class is going to be less of a hassle than I thought -- `__array_ufunc__` is definitely needed, though, so it will require numpy >=1.13. But given how nearly trivially the distributions could be used in `SkyCoord` -- modulo the annoyance of getting `erfa` routines to work (which can be solved; see discussion in #6865), it gives reasonable hope that making a masked `SkyCoord` will not be that hard.

taldcroft commented:
> First, I remain not sold on making Time mutable, and think that should be discussed separately.

It's hard to even start with the benefits, but immediately an example is an in-place sort, in-place arithmetic, table operations (which are a *lot* easier with mutability) and darn it just changing one value in a huge array because it needs to be different!  This happens all the time in real life.

But more to the point, why should astropy time be different from just about every other array data container in the Python universe (including pandas dates and numpy datetime which I presume to be mutable).  It was just an early (and what I expected to be temporary) implementation crutch to get the Time object out there and working.  But really the current state of immutability is all about technical implementation details that should not that difficult to iron out. 

> together with whether we want to (eventually) move that way for SkyCoord.

This is unrelated so can be discussed elsewhere.

>  Overall, my feeling is that abandoning immutability is in the end going to give much increased complexity and hurt performance.

Unlike coordinates, I think that Time is mostly done and stable.  What you are seeing in this PR *is* the increased complexity, or at least the vast majority.

mhvk commented:
@taldcroft - OK, some of those would indeed be quite nice to have... Though, e.g., the advantage of in-place sort, which can only really be done via an index array, is not as large as one might think. 

I guess a larger point on this is that if we go this route, we might as well try to make it as generic as possible. I don't think `__setitem__` can easily be done through `_apply`, but the same lists of attributes that can have array shapes and could be affected has to be consulted, and one can possibly pipe other in-place functions through a similar routine.

But really my main request is to separate out masks and mutability - they have very little to do with each other, and each is a large, involved change.

taldcroft commented:
> But really my main request is to separate out masks and mutability - they have very little to do with each other, and each is a large, involved change.

Having a `__setitem__` is a prerequisite for masking and that is the reason that the very first commit already has setting and masking intertwined.  In theory one can split this into a first PR that is setting-only, and then another one based on the setting PR which adds masking.  That would indeed be an ideal situation.  

I just don't see a really clean or quick way to do the separation.  Do you?  I have not done a census of all 25 commits to see if most of them are only setting or masking, but in any case it seems like at least 2 hours of work to get from here to there.  Where "there" is exactly "here" at the end, but in a different commit order.  So I'm a little not excited about doing this churn when time is limited.  But if you really need this separation then I will try.  I would be happy to put in github comments indicating "setitem" or "masking" for a non-obvious diff blocks to help out.

mhvk commented:
I guess I hadn't realized the masking and setting were so interrelated, in part as I had thought of masks more on the MaskedArray sense, where it isn't part of the data. And there would be little argument against changing a mask even if the object is immutable otherwise. But I guess with the current implementation that is not possible, since once you set a mask you are changing the data too.

Which perhaps suggests that splitting out is really not of the highest priority, but rather my other question about the implementation more generally. Would it not be better just on `Time` rather than the formats? And what is the benefit of using `nan` given the special casing with `jd2_filled`?

taldcroft commented:
About internal implementation:

- Using MaskedArray for jd1/jd2: here is what I wrote back in May:

"I first tried MaskedArray and got something mostly working (passing most but not all tests), but it ended up getting messy. In particular the interactions with Quantity got ugly because Quantity knows how to play with ndarray but not MaskedArray in arithmetic etc."

I honestly don't remember the details, but originally I was gung-ho on using MaskedArray but as I worked with the implementation, I ran into roadblocks that seemed like trouble.  If you want to play with this I have a branch I can push up.  Note however that this is at least as impacting on Format classes because you always need to get `self.jd1.data` instead of `self.jd1` for ERFA routines.  This would likely break existing non-astropy formats badly, instead of the current NaN implementation which will only break (for ERFA calls) if masks are set.  So basically no *existing* code will fail with this PR.  But we would recommend updating custom formats to use `jd2_filled` in calls to functions that cannot handle NaN.  I wonder if we could fix the upstream ERFA code to catch NaN and handle this?

Then of course there is performance since one has MaskedArray instead of ndarray everywhere, along with risk of more latent bugs in MaskedArray.

- Using NaN vs. a mask on Time

I had not really considered using a mask on Time before and my guess is that it would work OK but would have as many, though different, impacts on the code and a similar or greater increase in complexity / special-casing.  I think where you get hit in this approach is making sure that the mask is always handled correctly though any operations (copy, indexing, slicing, arithmetic, reshape, mean, reduction methods).  I'm not saying it is not possible to do all this and get it right, but I don't think the code will be substantially cleaner and more future-proof than the NaN implementation.  The exception is that existing custom Format classes would not need modification to handle masking.

With the NaN approach, it seems generally easier to get everything right.  Much of the above just comes for free because when you replicate or make a new Time object, the JD data (and embedded NaN's) are guaranteed to be correct, and the new object has nothing in cache and so the mask is guaranteed to be correct when requested.  Arithmetic just works, as do things like:
```
In [31]: t = Time([1,2,3,4], format='cxcsec')
In [33]: t[1] = np.ma.masked
In [34]: t.argsort()  # Missing values at the end was what I had expected, and it works!
Out[34]: array([0, 2, 3, 1])
```
Though it is not perfect, NaN is a reasonably good missing values sentinel.  So again, I don't doubt that this all can be done with a Time-level mask, but it does seem to be on the slippery slope of re-implementing a limited version of MaskedArray.  

Another advantage of the NaN approach is that you get support for serializing the mask completely for free!  I never even thought about this before and obviously didn't write any code to allow this.  It just works.  (Though to be fair ECSV does not work..)
```
In [39]: tm = Time([1,2,3,4], format='cxcsec')
In [40]: tm[1] = np.ma.masked
In [41]: t = Table([tm])
In [42]: t.write('junk.fits', overwrite=True)
In [43]: t2 = Table.read('junk.fits', astropy_native=True)
In [44]: t2
Out[44]: 
<Table length=4>
       col0       
      object      
------------------
 2450814.500011574
                --
 2450814.500034722
2450814.5000462965
```
On the whole, there is good reason that Pandas uses NaN within the array as a missing data sentinel and that a lot of numpy folks really wanted to do the same.

taldcroft commented:
@mhvk - I added documentation which explains potential code changes for custom Time formats, gives details of the missing values implementation, and provides a prominent warning that the implementation is provisional and might change.

This was marked as a high priority feature at the astropy 3.0 planning hangout and we'd really like to get it in this release.  An alternate implementation seems unrealistic at this point, but if we go in a different direction for 3.1 the impact will be quite small and people will have been warned.  We would also be able to deprecate but no remove `jd2_filled` and thus not actually break anything in 3.1.

mhvk commented:
@taldcroft - I'm very sorry but I don't see having the time to properly deal with this PR. I think I'm sold on the masking, but a bit less so on the mutability, in part because I feel fairly strongly that `Time` and `SkyCoord` should behave similarly. For both, though, I have concerns about the implementation:
1. For masking, I think we should *not* use a NaN marker, but rather have a separate array, so that one can unmask/recover if needed. In part I think this because with a separate array I can see how to implement the same for `SkyCoord` (and maybe `Quantity`, though that is perhaps harder; with `__array_ufunc__`, though, it has become decidedly simpler, and in 3.1 we can require numpy 1.13 and thus count on it). I also am fairly confident that it would be reasonably easy to do given the work you have already, and leaving the format classes untouched. It may even be possible to have it as a `Masked` mixin, similar to `ShapedLikeNDArray` and the new `Distribution` that @eteq and I worked on (though that mixin is secondary and definitely can be done later; it is mostly that I see a path towards that with a separate array, and much less so with the NaN approach).
2. For setting items, I mostly would like to have setting work all the time, i.e., independent of what the location is. I think that would automatically lead to an implementation that could be moved to `SkyCoord` as well. Furthermore, I want to think through better the implications; e.g., some of the broadcasting presently done may have to be changed.

I realize this is rather annoying, since you had versions of this already for a long time; unfortunately, much of my astropy time has gone into coordinates (but with results I'm very happy with). Unless there is someone else who does have time to thoroughly review this now -- it would be great to have someone else who understood `Time` well! -- my suggestion would be to *not* include it in 3.0, but rather aim for 3.1. That said, to be sure that it doesn't get further delay, I'm happy to merge soon into 3.1 development, so further work can be based on top of it (which doesn't have to care as much about the implementation).

taldcroft commented:
Humm, it seems we are at a bit of an impasse.

> I feel fairly strongly that Time and SkyCoord should behave similarly.

Can you elaborate on why that is?  I feel strongly that Time should behave like a standard Table column and be able to fully participate in table operations.  This is exactly what this PR and #6888 do, and I would argue this is more compelling than Time and SkyCoord behaving similarly.  Note also that in the past @eteq has expressed very strong reservations about making coordinates mutable, so personally I haven't had an expectation that will happen any time soon.

> For masking, I think we should not use a NaN marker, but rather have a separate array, so that one can unmask/recover if needed.

I don't see this as a requirement.  I think that what we need to implement is an indicator of *missing or invalid data*, NOT *masked* data that can or even should be recovered.  At least in my experience within astropy Tables and in SQL databases, missing data are really missing.  I would take Pandas as a strong community endorsement of both this philosophical treatment of missing data and the implementation of using NaN as a marker (see http://pandas.pydata.org/pandas-docs/version/0.20/missing_data.html).    There is no "recovering" missing data in Pandas.

Coming back to practical concerns for an important use case: using NaN for Time turns out to be very convenient because we can write a Time object with missing data to FITS, with absolutely no extra effort.  This works right now.  Using a separate mask attribute gets us into a bit of a bind because the Time FITS standard (and code) does not support this.  So Time objects would need to be quasi-serialized, which is to say the mask would need to be extracted and stored in the table on the side.  Everything is possible, but this does not fit into our current machinery which explicitly ignores Time columns in the table serialize process for FITS.

> For setting items, I mostly would like to have setting work all the time, i.e., independent of what the location is.

This is purely a limitation of Time in that we cannot easily transform a Time object from one location to another.  If that is addressed then setitem could do that as well.  But that is actually missing a bigger point, namely that the current implementation works for the overwhelming majority of real world use cases.  Blocking this because it raises a (nice informative) exception for an obscure corner case doesn't make sense.

eteq commented:
A few quick thoughts:

1) In *general* tend to agree with @mhvk on the overall philosophy of preferring a mask array over NaN.  I think the fact that Pandas does this is actually a good reason for why I *prefer* Table over Pandas. "masked" is not always "missing", nor is it always "not a number".  That is particularly true in astronomy where "missing" sometimes means "cannot be recovered" and other times means "problematic but there's at least some value to the number here".  Also integers can't be NaNs. All that said, I see how in the *specific* case of Time it's less clear that's important - they aren't integers anyway, and I can't think of any obvious case where you know the time but want to mask it.

2) I'm not certain whether `SkyCoord` and `Time` really need to have the same mutability behavior.  My concerns about making `SkyCoord` mutable are based on the fact that frames can contain a *lot* of complex information in them that has to be managed in a way that keeps them consistent when modified. `Time` is a rather simpler beast, with basically only the two jds, the format, scale, and the location to worry about (and I even think the location really shouldn't be considered...).  So I'm less certain that @mhvk seems to be that we can make an implementation work for both.

So I tend to think this is OK as it stands philisophically... but I'm not sure what to do about the "who can review it" problem.

astrofrog commented on astropy/time/core.py:
```diff
@@ -1013,9 +1135,17 @@ def argmax(self, axis=None, out=None):
+            if axis is not None:
+                approx = np.expand_dims(approx, axis)
+        else:
+            approx = np.max(jd, axis, keepdims=True)
```
Note that for 3.1 we are dropping Numpy < 1.13 so these kinds of blocks can be simplified. However https://github.com/astropy/astropy/pull/7058 will need a little work before being merged, so we can always just merge this as-is then remove later.

astrofrog commented on astropy/time/formats.py:
```diff
@@ -141,11 +142,42 @@ def scale(self):
-                (val2 is None or
-                 val2.dtype == np.double and np.all(np.isfinite(val2)))):
+        # val1 cannot contain nan, but val2 can contain nan
+        ok1 = val1.dtype == np.double and np.all(np.isfinite(val1))
```
Why ``isfinite`` and not ``isnan``? This will catch inf/-inf

taldcroft replied:
This line for `ok1` is the same as it has always been and says that the user input must be double and finite and not NaN.  (`isfinite(nan) is False`).

The next line for `ok2` is mostly the same but requires that all values be a finite number or NaN.  (`isinf(nan) is False`).

Summary: same as before except that NaN is allowed for `val2`.

astrofrog replied:
Ok sounds good!

astrofrog commented on astropy/time/tests/test_basic.py:
```diff
@@ -1168,3 +1167,138 @@ def test_bytes_input():
+    # Fails because the right hand side has location=None
+    with pytest.raises(ValueError) as err:
+        t[0, 0] = Time(-1, format='cxcsec')
+    assert 'cannot set to Time with different location' in str(err)
```
The error message doesn't seem quite right here - isn't the issue more that on the right the location isn't set so it's ambiguous?

taldcroft replied:
It's different in the sense that the left side has `location = EarthLocation(...)` while the right side has `location = None`.  I think the exception will let the user diagnose the problem by hinting to print the location of left and right sides.  I could make the exception message be `.. different location attribute` to be even more clear?  Or something else?

astrofrog replied:
Yes, saying different location attribute - maybe could you explicitly say what the location is in each case? (expected ..., got ...)

astrofrog commented on docs/time/index.rst:
```diff
@@ -576,14 +586,127 @@ The two should be very close to each other.
+
+  >>> t.mask
+  array([False, False,  True, False]...)
+  >>> t[:2] = np.ma.masked
```
I think you should explain that and why it's not possible to set the mask using e.g. t.mask[:2] = True

taldcroft replied:
Will do.

taldcroft commented:
The syntax of `t.mask[1] = True` is trivial, but of course the data behind the mask are lost and you cannot unmask.  Is there an obvious use-case for masking then unmasking Time?  I note that numpy MaskedArray discourages manipulation of the mask directly as you showed, and I've gone in that direction.

astrofrog commented:
> The syntax of t.mask[1] = True is trivial, but of course the data behind the mask are lost and you cannot unmask. Is there an obvious use-case for masking then unmasking Time? I note that numpy MaskedArray discourages manipulation of the mask directly as you showed, and I've gone in that direction.

Right, I think that given the NaN-based implementation we shouldn't allow this since unmasking doesn't work. I was not aware that Numpy were moving away from that. In any case, my point is that ``t.mask[1] = True`` isn't possible anyway right now, so we're not taking away something users can do already. I think we should discuss whether we want to move to something that allows that after this PR.

eteq commented:
Realizing I've had several out-of-band conversations on this without commenting in this PR... I haven't been able to review in detail but I have some general thoughts.  To summarize, now that @astrofrog clarified that the NaN part was implementation and *not* visible to the user, I'm fine with this.  I'm not convinced we don't want mutable masks in a general sense (e.g. one could argue this might be good in `SkyCoord`). And I agree with @mhvk that we want similar behavior as much as possible across packages like `time`, `table`, and `coordinates`.  But I think it's safe to *start* with an immutible mask (which my understanding is that this does), and potentially make it mutable *later* if we think that's reasonable across the packages.  That's safer than the converse, so I'm OK with it.


> I note that numpy MaskedArray discourages manipulation of the mask directly as you showed

I didn't appreciate that either... I'm also not sure that I agree - I think one of the *advantages* of table is that it allows this.  But I think it's a moot point for *this* PR (see above).

mhvk commented:
>  I note that numpy MaskedArray discourages manipulation of the mask directly as you showed

I don't know that this is true. It can just be messy for arrays with a structured dtype (which part does one mask).

taldcroft commented:
@mhvk - agreed it is not critical to this discussion nor necessarily the viewpoint of astropy, but I stand by my statement.  :smile:

From https://docs.scipy.org/doc/numpy-1.13.0/reference/maskedarray.generic.html#modifying-the-mask:

![image](https://user-images.githubusercontent.com/348089/35752295-6358c086-0829-11e8-9517-2a5adf0a5030.png)

mhvk commented:
I agree with you (and the documentation) that it is better to just write `masked` to an element. I guess my main point, rather indirectly, is that `MaskedArray` does its utmost to preserve the masked values (arguably too much effort!) and that this PR breaks that expectation. Anyway, not directly relevant indeed.

taldcroft commented:
With all due respect for the folks that have worked hard on MaskedArray, I just can't hold it up as an exemplar of code and interface that astropy should strive toward,  This is numpy 1.14.
```
In [2]: x = np.ma.array([1,0])

In [3]: y = np.ma.array([1,1])

In [4]: y/x
Out[4]: 
masked_array(data=[1.0, --],
             mask=[False,  True],
       fill_value=1e+20)

In [5]: z = y/x

In [6]: z.mask[1] = False

In [7]: z
Out[7]: 
masked_array(data=[1.0, 1.0],
             mask=[False, False],
       fill_value=1e+20)
```

taldcroft commented:
@astrofrog - I think I addressed your comments.

![image](https://user-images.githubusercontent.com/348089/35770910-818c2948-08f2-11e8-905e-89d0dfa86fe9.png)

Thanks to all for your reviews!

astrofrog commented:
@taldcroft - looks good! This is good to go :)
